### ISSUE TITLE
Docker stats for all container

### ISSUE THREAD

**Issue by jeanpralo** [docker/docker#22052]
Hi there,

Do you guys think it could be a good idea to have in the API an endpoint which allows to get the stats of all the running containers at once?

At the moment I am using this collectd plugin (https://github.com/lebauce/docker-collectd-plugin) to get the stats out of my server. 

The problem I am seeing is that when I am running 100+ container the server starts loading because I am making too many requests too often to the docker daemon (aka one call per container to the /stats endpoint every 20s).

If we could get all the stats at once that would fix this problem, for now I have implemented some rate limit on that plugin but not sure if it is the best as I do  not have a real time info anymore.

Cheers.

**Comment by thaJeztah** [comment]
So basically, move this functionality: https://github.com/docker/docker/pull/16742 server / daemon side?

@runcom wdyt? I don't know how this will interact with (e.g.) a Swarm setup

**Comment by jeanpralo** [comment]
@thaJeztah  yeah if that's possible that would be great otherwise the more containers you run the less scalable it becomes.

**Comment by icecrime** [comment]
I think it makes sense (:wine_glass: @jeanpralo)

**Comment by WeiZhang555** [comment]
I like this idea, it would save lots of TCP connections, but maybe it will impact the Swarm functions as @thaJeztah mentioned.

**Comment by runcom** [comment]
makes sense :) at the beginning it was just to painful to think about exposing an API to collect all stats so I went the client way

**Comment by thaJeztah** [comment]
ping @vieux @dongluochen any thoughts on implications for Swarm?

**Comment by bprodoehl** [comment]
I'd also like to see this, but I think the high CPU load in #21837 would probably need to be addressed.  It seems to take a lot of effort for the daemon to get the stats together in a format that it can spit out, and I think an API endpoint that collects stats on all containers would really expose that.

**Comment by vincentwoo** [comment]
I think this is a must-have for the stats API surface. Anyone who uses the stats endpoints almost certainly uses it for all of their containers. Having the stats streams per-container is very awkward. You have to continually track which containers you have subscribed to, etc.

**Comment by thaJeztah** [comment]
If someone wants to make a pull request for this; based on the discussion above, I think it will be accepted

**Comment by vincentwoo** [comment]
@rileysaurus if you're looking for something to do, this issue is a decent candidate. we might even be able to get a twofer addressing https://github.com/docker/docker/issues/21837, which also seems pretty nasty.

**Comment by WeiZhang555** [comment]
I'd like to take care of this if no one else is already working on this.

So #dibs  :smile:

**Comment by vincentwoo** [comment]
don't let me down, wei

**Comment by WeiZhang555** [comment]
@vincentwoo Oh, so much pressure :laughing:  I'll try my best :guitar:

**Comment by jeanpralo** [comment]
Sad end for that first PR :( but thanks for the work @WeiZhang555  


@thaJeztah @icecrime  any chance we can raise a priority on that issue. Getting stats out of docker is something that one would expect.  At the moment servers with hundred of containers are not really go at doing that :'(

**Comment by thaJeztah** [comment]
Focus will most likely be on https://github.com/docker/docker/issues/27307, but input is welcome there 👍

**Comment by aryamsft** [comment]
Has this feature been implemented?  I'm looking here:  https://docs.docker.com/engine/api/v1.40/#operation/ContainerStats  but I don't see a way to get all the stats at once.

**Comment by cpuguy83** [comment]
@thaJeztah WDYT if we did something like `GET /containers/json?fetchContainerStats=true` and cram in the stats for each container listed?
This gets around issues with polluting the API path with weird things and solves issues for people who just want to get a list for all containers... and it also happens to support filtering and such already.

**Comment by cpuguy83** [comment]
As I recall we already have a `size` param to fetch the container rootfs size as well, so not really out of left field.

**Comment by thaJeztah** [comment]
I think we had a PR implementing this at some point, but it was rejected at the time (will have to check)

**Comment by aryamsft** [comment]
Hey @thaJeztah any update on this?

**Comment by thaJeztah** [comment]
So, the old PR was https://github.com/moby/moby/pull/25361, which has a lot of discussion (some of the comments are hidden by GitHub, so you might have to click to get those shown)

**Comment by aryamsft** [comment]
I read through the thread, seems like there is a lot of contention around the restful API design of it (the naming of the endpoint).  But, that thread ended 2 years 8 months ago.  Can this problem be easily solved now or does the contention still exist?

**Comment by aryamsft** [comment]
Hi, bumping this thread.

@thaJeztah Was there any traction on this issue?  There is a feature I'm working on where I need to pull stats for all the containers.  I would rather not make a restful call for each and every container as there could be hundreds of containers running on a particular machine.  

Can we add a parameter to the restful call for pulling all container stats (or adding a new restful api)?  I'm sure this would be useful for other people as I don't think I'm the only person running into this issue.

Thanks!

**Comment by cpuguy83** [comment]
I think adding a flag like `GET /containers/json?fetchContainerStats=true`​​​​​​​ is the best bet. It is a similar thing for `docker service ps` where we have a flag to fetch all the task states rather than dealing with N+1 issues.

We also have a similar flag on the same containers endpoint for fetching the RW layer size. It fits in well.

@thaJeztah WDYT?

**Comment by aryamsft** [comment]
@thaJeztah ping on this?

**Comment by thaJeztah** [comment]
> I think adding a flag like `GET /containers/json?fetchContainerStats=true​​​​​`​​ is the best bet.

Bit on the fence on that suggestion; stats are collected through a different endpoint (`/containers/{name:.*}/stats`);

- collecting `stats` is already a heavy operation; wondering if adding the information of `/containers/json` to it would make it even more heavy
- the `/containers/{name:.*}/stats` endpoint allows _streaming_ the output (refreshing the stats on the daemon side, and stream the results); this would have to be implemented for that endpoint as well; does that mean `?stream=true` would also have to be supported when listing containers without stats?

I think reasons above were why the previous PR converged to using a placeholder `-` as container-id, so `/containers/-/stats` would fetch all containers (`-` is not a valid container-id, hence wouldn't be able to conflict with a container named `-`), whereas `/containers/{name|id}/stats` would fetch stats for a single container.

That said; an endpoint that would default to _all_ containers would more easily allow adding (e.g.) `--filter` or `--limit` (show stats for all containers matching certain criteria)

**Comment by cpuguy83** [comment]
> wondering if adding the information of /containers/json to it would make it even more heavy

Not really, not much.

>  this would have to be implemented for that endpoint as well

No this would be for 1-off requests.
Really we probably don't need the streaming thing to begin with, clients can just request again and optimize the transport.

> That said; an endpoint that would default to all containers would more easily allow adding (e.g.) --filter or --limit (show stats for all containers matching certain criteria)

The nice thing about having it on the list endpoint is we already have all this. We don't need to create a new endpoint, just add new data to the existing one.

**Comment by tianon** [comment]
IIRC, even for a one-shot it has to monitor for some period to measure
change over time right? (ISTR asking for one instead of a stream still
takes ~a full second or so)

**Comment by cpuguy83** [comment]
Yes it does it over a 1s interval. There are few things happening that slow this down.
But definitely waiting for CPU over at least 1s is the majority of the time.

I think what @aryamsft wants to avoid is First getting the list of containers and then issuing requests for each one.
As is it would still take time to actually fetch the stats.

**Comment by cpuguy83** [comment]
Does it make sense to provide an API that lets the client calculate CPU usage?

**Comment by cpuguy83** [comment]
Actually I'm wondering why we are priming CPU stats for `stream=false`, the CLI could just get two frames expecting the daemon to handle it.

**Comment by aryamsft** [comment]
So for my scenario (and I'm sure other people's scenarios) is simply getting the stats for all the containers in a single call.  If there is a node/VM that has 100+ containers running, making 100 calls and storing/parsing/massaging each return value into some list is a high amount of work.  Making a single call, then storing/parsing/massaging all the data at once is far simpler.  I don't really need any streaming type functionality.  If someone needs to get a fresh set of data they can simply use existing API's and make the call again every few seconds.

I'm ok with whatever design you guys come up with since this is your domain, as long as it is easy and obvious to use :)

**Comment by cpuguy83** [comment]
@aryamsft Understood. I've moved my discussion on fixing the slowness of the API call to https://github.com/moby/moby/issues/40437

Thanks!
